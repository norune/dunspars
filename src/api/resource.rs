use std::collections::HashMap;

use anyhow::{anyhow, bail, Result};
use futures::stream::FuturesOrdered;
use futures::StreamExt;
use regex::Regex;

use rustemon::client::RustemonClient;
use rustemon::games::version_group as rustemon_version;

use super::utils;
use crate::pokemon::Game;

pub enum ResourceResult {
    Valid,
    Invalid(Vec<String>),
}

#[allow(async_fn_in_trait)]
pub trait Resource: Sized {
    fn get_matches(&self, value: &str) -> Vec<String> {
        self.resource()
            .iter()
            .filter_map(|r| {
                let close_enough = if !r.is_empty() && !value.is_empty() {
                    let first_r = r.chars().next().unwrap();
                    let first_value = value.chars().next().unwrap();

                    // Only perform spellcheck on first character match; potentially expensive
                    first_r == first_value && strsim::levenshtein(r, value) < 4
                } else {
                    false
                };

                if r.contains(value) || close_enough {
                    Some(r.clone())
                } else {
                    None
                }
            })
            .collect::<Vec<String>>()
    }

    fn check(&self, value: &str) -> ResourceResult {
        let matches = self.get_matches(value);
        if matches.iter().any(|m| *m == value) {
            ResourceResult::Valid
        } else {
            ResourceResult::Invalid(matches)
        }
    }

    fn validate(&self, value: &str) -> Result<String> {
        let value = value.to_lowercase();
        match self.check(&value) {
            ResourceResult::Valid => Ok(value),
            ResourceResult::Invalid(matches) => bail!(Self::invalid_message(&value, &matches)),
        }
    }

    fn invalid_message(value: &str, matches: &[String]) -> String {
        let resource_name = Self::label();
        let mut message = format!("{resource_name} '{value}' not found.");

        if matches.len() > 20 {
            message += " Potential matches found; too many to display.";
        } else if !matches.is_empty() {
            message += &format!(" Potential matches: {}.", matches.join(" "));
        }

        message
    }

    async fn try_new(api: &RustemonClient) -> Result<Self>;
    fn resource(&self) -> Vec<String>;
    fn label() -> &'static str;
}

#[derive(Debug)]
pub struct PokemonResource {
    resource: Vec<String>,
}
impl Resource for PokemonResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_pokemon(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Pokémon"
    }
}

#[derive(Debug)]
pub struct TypeResource {
    resource: Vec<String>,
}
impl Resource for TypeResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_types(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Type"
    }
}

#[derive(Debug)]
pub struct MoveResource {
    resource: Vec<String>,
}
impl Resource for MoveResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_moves(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Move"
    }
}

#[derive(Debug)]
pub struct AbilityResource {
    resource: Vec<String>,
}
impl Resource for AbilityResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_abilities(client).await?;
        Ok(Self { resource })
    }

    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Ability"
    }
}

#[derive(Debug)]
pub struct GameResource {
    resource: HashMap<String, Game>,
    gen_url_regex: Regex,
}
impl Resource for GameResource {
    async fn try_new(client: &RustemonClient) -> Result<Self> {
        // PokéAPI keeps generation names in Roman numerals.
        // Might be quicker to just take it from resource urls via regex instead.
        // Regex compilation is expensive, so we're compiling it just once here.
        let gen_url_regex = Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap();

        let mut resource = HashMap::new();
        let game_names = utils::get_all_games(client).await?;
        let game_data_futures: FuturesOrdered<_> = game_names
            .iter()
            .map(|g| rustemon_version::get_by_name(g, client))
            .collect();
        let game_data: Vec<_> = game_data_futures.collect().await;

        for (i, game) in game_data.into_iter().enumerate() {
            let game = game?;
            let generation = capture_gen_url(&game.generation.url, &gen_url_regex).unwrap();
            resource.insert(game.name.clone(), Game::new(game.name, i as u8, generation));
        }

        Ok(Self {
            resource,
            gen_url_regex,
        })
    }

    fn resource(&self) -> Vec<String> {
        let mut games = self.resource.iter().map(|r| r.1).collect::<Vec<&Game>>();
        games.sort_by_key(|g| g.order);

        games
            .iter()
            .map(|g| g.name.clone())
            .collect::<Vec<String>>()
    }

    fn label() -> &'static str {
        "Game"
    }
}
impl GameResource {
    pub fn get_gen(&self, game: &str) -> u8 {
        self.resource.get(game).unwrap().generation
    }

    pub fn get_gen_from_url(&self, url: &str) -> u8 {
        capture_gen_url(url, &self.gen_url_regex).unwrap()
    }
}

fn capture_gen_url(url: &str, gen_url_regex: &Regex) -> Result<u8> {
    if let Some(caps) = gen_url_regex.captures(url) {
        Ok(caps["gen"].parse::<u8>()?)
    } else {
        Err(anyhow!("Generation not found in resource url"))
    }
}
