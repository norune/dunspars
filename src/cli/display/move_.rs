use super::{Colors, DisplayComponent};
use crate::pokemon::Move;

use std::fmt;

use indoc::writedoc;

impl fmt::Display for DisplayComponent<&Move<'_>> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        let Move {
            power,
            accuracy,
            pp,
            name,
            effect,
            damage_class,
            effect_chance,
            type_,
            ..
        } = self.context;

        let power = if let Some(power) = power {
            power.to_string()
        } else {
            "N/A".to_string()
        };
        let accuracy = if let Some(accuracy) = accuracy {
            accuracy.to_string()
        } else {
            "N/A".to_string()
        };
        let pp = if let Some(pp) = pp {
            pp.to_string()
        } else {
            "N/A".to_string()
        };

        let stats = format!(
            "power: {red}{power:3}{red:#}  accuracy: {green}{accuracy:3}{green:#}  pp: {blue}{pp:3}{blue:#}",
            red = self.ansi(Colors::Red),
            green = self.ansi(Colors::Green),
            blue = self.ansi(Colors::Blue),
        );

        let effect_text = if let Some(chance) = effect_chance {
            effect.replace("$effect_chance", &chance.to_string())
        } else {
            effect.to_string()
        };

        writedoc! {
            f,
            "{header}{name}{header:#}
            {type_} {damage_class}
            {stats}
            {effect_text}",
            header = self.ansi_bold(Colors::Header)
        }
    }
}
