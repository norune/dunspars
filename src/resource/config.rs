use super::{app_config_directory, AppFile, YamlFile};

use std::collections::HashMap;
use std::path::PathBuf;

pub struct ConfigFile {
    path: PathBuf,
}
impl ConfigFile {
    pub fn new(path: PathBuf) -> Self {
        Self { path }
    }
}
impl AppFile for ConfigFile {
    fn path(&self) -> &PathBuf {
        &self.path
    }
}
impl YamlFile for ConfigFile {
    type YamlData = ConfigCollection;
}
impl Default for ConfigFile {
    fn default() -> Self {
        Self::new(app_config_directory("config.yaml"))
    }
}

#[derive(Default, Debug, serde::Serialize, serde::Deserialize)]
pub struct ConfigCollection {
    config: HashMap<String, String>,
}
impl ConfigCollection {
    pub fn get_collection(&self) -> &HashMap<String, String> {
        &self.config
    }

    pub fn get_value(&self, key: &str) -> Option<&String> {
        self.config.get(key)
    }

    pub fn set_value(&mut self, key: &str, value: &str) -> Option<String> {
        self.config.insert(String::from(key), String::from(value))
    }

    pub fn unset_value(&mut self, key: &str) -> Option<String> {
        self.config.remove(key)
    }
}
