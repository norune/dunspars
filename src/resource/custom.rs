use super::{app_config_directory, AppFile, YamlFile};

use std::path::PathBuf;

#[derive(serde::Serialize, serde::Deserialize)]
pub struct CustomFile {
    path: PathBuf,
}
impl CustomFile {
    pub fn new(path: PathBuf) -> Self {
        Self { path }
    }
}
impl AppFile for CustomFile {
    fn path(&self) -> &PathBuf {
        &self.path
    }
}
impl YamlFile for CustomFile {
    type YamlData = CustomCollection;
}
impl Default for CustomFile {
    fn default() -> Self {
        Self::new(app_config_directory("custom.yaml"))
    }
}

#[derive(Default, Debug, serde::Serialize, serde::Deserialize)]
pub struct CustomCollection {
    pokemon: Vec<CustomPokemon>,
}
impl CustomCollection {
    pub fn find_pokemon(&self, nickname: &str) -> Option<&CustomPokemon> {
        self.pokemon
            .iter()
            .find(|p| p.nickname.to_lowercase() == nickname.to_lowercase())
    }
}

#[derive(Debug, serde::Serialize, serde::Deserialize)]
pub struct CustomPokemon {
    pub nickname: String,
    pub base: String,
    pub generation: u8,
    pub moves: Vec<String>,
    pub types: Option<(String, Option<String>)>,
}
