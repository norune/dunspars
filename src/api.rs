use std::collections::HashMap;

use anyhow::{bail, Result};
use regex::Regex;

use dirs;

use rustemon::client::{CACacheManager, RustemonClient, RustemonClientBuilder};
use rustemon::games::version_group as rustemon_version;
use rustemon::moves::move_ as rustemon_moves;
use rustemon::pokemon::pokemon_species as rustemon_species;
use rustemon::pokemon::{
    ability as rustemon_ability, pokemon as rustemon_pokemon, type_ as rustemon_type,
};
use rustemon::Follow;

use rustemon::model::evolution::{
    ChainLink as RustemonEvoStep, EvolutionChain as RustemonEvoRoot,
    EvolutionDetail as RustemonEvoMethod,
};
use rustemon::model::games::VersionGroup as RustemonVersion;
use rustemon::model::moves::{Move as RustemonMove, PastMoveStatValues as RustemonPastMoveStats};
use rustemon::model::pokemon::{
    Ability as RustemonAbility, AbilityEffectChange as RustemonPastAbilityEffect,
    Pokemon as RustemonPokemon, PokemonStat as RustemonStat, PokemonType as RustemonTypeSlot,
    PokemonTypePast as RustemonPastPokemonType, Type as RustemonType,
    TypeRelations as RustemonTypeRelations, TypeRelationsPast as RustemonPastTypeRelations,
};
use rustemon::model::resource::{Effect as RustemonEffect, VerboseEffect as RustemonVerboseEffect};

use crate::pokemon::{
    Ability, EvolutionMethod, EvolutionStep, Move, PokemonData, Stats, Type, TypeChart,
};

pub struct ApiWrapper {
    client: RustemonClient,
}

impl ApiWrapper {
    pub fn try_new() -> Result<Self> {
        let mut client = RustemonClientBuilder::default();
        let mut cache_dir = if let Some(home_dir) = dirs::home_dir() {
            home_dir
        } else {
            bail!("Home directory not found")
        };
        cache_dir.push(".cache/dunspars/rustemon");

        let cache_manager = CACacheManager { path: cache_dir };
        client.with_manager(cache_manager);

        Ok(Self {
            client: client.try_build()?,
        })
    }
}

impl ApiWrapper {
    pub async fn get_pokemon(&self, pokemon: &str, game: &str) -> Result<PokemonData> {
        let RustemonPokemon {
            name,
            types,
            past_types,
            moves,
            stats,
            abilities,
            species,
            ..
        } = rustemon_pokemon::get_by_name(pokemon, &self.client).await?;
        let generation = self.get_generation(game).await?;

        let pokemon_types = self
            .match_past(generation, past_types)
            .await
            .unwrap_or(types);
        let primary_type = pokemon_types
            .iter()
            .find(|t| t.slot == 1)
            .unwrap()
            .type_
            .name
            .clone();
        let secondary_type = pokemon_types
            .iter()
            .find(|t| t.slot == 2)
            .map(|t| t.type_.name.clone());

        let mut learn_moves = HashMap::new();
        moves.iter().for_each(|mv| {
            let version_group = mv
                .version_group_details
                .iter()
                .find(|vg| vg.version_group.name == game);
            if let Some(vg) = version_group {
                learn_moves.insert(
                    mv.move_.name.clone(),
                    (vg.move_learn_method.name.clone(), vg.level_learned_at),
                );
            }
        });

        let abilities = abilities
            .iter()
            .map(|a| (a.ability.name.clone(), a.is_hidden))
            .collect::<Vec<_>>();

        Ok(PokemonData {
            name,
            primary_type,
            secondary_type,
            learn_moves,
            abilities,
            species: species.name,
            stats: extract_stats(stats),
            game: game.to_string(),
            generation,
            api: self,
        })
    }

    pub async fn get_type(&self, type_str: &str, generation: u8) -> Result<Type> {
        let RustemonType {
            name,
            damage_relations,
            past_damage_relations,
            ..
        } = rustemon_type::get_by_name(type_str, &self.client).await?;

        let relations = self
            .match_past(generation, past_damage_relations)
            .await
            .unwrap_or(damage_relations);

        let mut offense_chart = HashMap::new();
        let mut defense_chart = HashMap::new();

        relations.no_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 0.0);
        });
        relations.half_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 0.5);
        });
        relations.double_damage_to.iter().for_each(|t| {
            offense_chart.insert(t.name.to_string(), 2.0);
        });

        relations.no_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 0.0);
        });
        relations.half_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 0.5);
        });
        relations.double_damage_from.iter().for_each(|t| {
            defense_chart.insert(t.name.to_string(), 2.0);
        });

        Ok(Type {
            name,
            offense_chart: TypeChart::new(offense_chart),
            defense_chart: TypeChart::new(defense_chart),
            generation,
            api: self,
        })
    }

    pub async fn get_move(&self, name: &str, generation: u8) -> Result<Move> {
        let RustemonMove {
            name,
            mut accuracy,
            mut power,
            mut pp,
            damage_class,
            mut type_,
            mut effect_chance,
            effect_entries,
            effect_changes,
            past_values,
            ..
        } = rustemon_moves::get_by_name(name, &self.client).await?;

        let RustemonVerboseEffect {
            mut effect,
            mut short_effect,
            ..
        } = effect_entries
            .into_iter()
            .find(|e| e.language.name == "en")
            .unwrap_or_default();

        if let Some(past_stats) = self.match_past(generation, past_values).await {
            accuracy = past_stats.accuracy.or(accuracy);
            power = past_stats.power.or(power);
            pp = past_stats.pp.or(pp);
            effect_chance = past_stats.effect_chance.or(effect_chance);

            if let Some(t) = past_stats.type_ {
                type_ = t;
            }

            if let Some(entry) = past_stats
                .effect_entries
                .into_iter()
                .find(|e| e.language.name == "en")
            {
                effect = entry.effect;
                short_effect = entry.short_effect;
            }
        }

        if let Some(past_effects) = self.match_past(generation, effect_changes).await {
            if let Some(past_effect) = past_effects.into_iter().find(|e| e.language.name == "en") {
                effect += format!("\n\nGeneration {generation}: {}", past_effect.effect).as_str();
            }
        }

        Ok(Move {
            name,
            accuracy,
            power,
            pp,
            damage_class: damage_class.name,
            type_: type_.name,
            effect_chance,
            effect,
            short_effect,
            generation,
            api: self,
        })
    }

    pub async fn get_ability(&self, name: &str, generation: u8) -> Result<Ability> {
        let RustemonAbility {
            name,
            effect_entries,
            effect_changes,
            ..
        } = rustemon_ability::get_by_name(name, &self.client).await?;

        let RustemonVerboseEffect {
            mut effect,
            short_effect,
            ..
        } = effect_entries
            .into_iter()
            .find(|e| e.language.name == "en")
            .unwrap_or_default();

        if let Some(past_effects) = self.match_past(generation, effect_changes).await {
            if let Some(past_effect) = past_effects.into_iter().find(|e| e.language.name == "en") {
                effect += format!("\n\nGeneration {generation}: {}", past_effect.effect).as_str();
            }
        }

        Ok(Ability {
            name,
            effect,
            short_effect,
            generation,
            api: self,
        })
    }

    pub async fn get_generation(&self, game: &str) -> Result<u8> {
        let RustemonVersion { generation, .. } =
            rustemon_version::get_by_name(game, &self.client).await?;

        extract_gen_from_url(&generation.url)
    }

    pub async fn get_evolution_steps(&self, species: &str) -> Result<EvolutionStep> {
        let RustemonEvoRoot { chain, .. } = rustemon_species::get_by_name(species, &self.client)
            .await?
            .evolution_chain
            .unwrap()
            .follow(&self.client)
            .await?;
        let evolution_step = traverse_chain(chain);

        Ok(evolution_step)
    }

    pub async fn get_all_moves(&self) -> Result<Vec<String>> {
        Ok(rustemon_moves::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    pub async fn get_all_abilities(&self) -> Result<Vec<String>> {
        Ok(rustemon_ability::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }
    pub async fn get_all_games(&self) -> Result<Vec<String>> {
        Ok(rustemon_version::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    pub async fn get_all_types(&self) -> Result<Vec<String>> {
        Ok(rustemon_type::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }
    pub async fn get_all_pokemon(&self) -> Result<Vec<String>> {
        Ok(rustemon_pokemon::get_all_entries(&self.client)
            .await?
            .into_iter()
            .map(|p| p.name)
            .collect::<Vec<String>>())
    }

    async fn match_past<T: Past<U>, U>(&self, current_generation: u8, pasts: Vec<T>) -> Option<U> {
        let mut oldest_value = None;
        let mut oldest_generation = 255;

        for past in pasts {
            let past_generation = past.generation(self).await;
            if current_generation <= past_generation && past_generation <= oldest_generation {
                oldest_value = Some(past.value());
                oldest_generation = past_generation;
            }
        }

        oldest_value
    }
}

fn traverse_chain(chain_link: RustemonEvoStep) -> EvolutionStep {
    let evolution_methods = chain_link
        .evolution_details
        .into_iter()
        .map(convert_to_evolution_method)
        .collect();

    if !chain_link.evolves_to.is_empty() {
        let evolves_to = chain_link
            .evolves_to
            .into_iter()
            .map(traverse_chain)
            .collect();

        EvolutionStep::new(chain_link.species.name, evolution_methods, evolves_to)
    } else {
        EvolutionStep::new(chain_link.species.name, evolution_methods, vec![])
    }
}

fn convert_to_evolution_method(evolution: RustemonEvoMethod) -> EvolutionMethod {
    let mut method = EvolutionMethod::new(evolution.trigger.name);
    if let Some(item) = evolution.item {
        method = method.item(item.name);
    }
    if let Some(gender) = evolution.gender {
        method = method.gender(gender);
    }
    if let Some(held_item) = evolution.held_item {
        method = method.held_item(held_item.name);
    }
    if let Some(known_move) = evolution.known_move {
        method = method.known_move(known_move.name);
    }
    if let Some(known_move_type) = evolution.known_move_type {
        method = method.known_move_type(known_move_type.name);
    }
    if let Some(location) = evolution.location {
        method = method.location(location.name);
    }
    if let Some(min_level) = evolution.min_level {
        method = method.min_level(min_level);
    }
    if let Some(min_happiness) = evolution.min_happiness {
        method = method.min_happiness(min_happiness);
    }
    if let Some(min_beauty) = evolution.min_beauty {
        method = method.min_beauty(min_beauty);
    }
    if let Some(min_affection) = evolution.min_affection {
        method = method.min_affection(min_affection);
    }
    if let Some(party_species) = evolution.party_species {
        method = method.party_species(party_species.name);
    }
    if let Some(party_type) = evolution.party_type {
        method = method.party_type(party_type.name);
    }
    if let Some(relative_physical_stats) = evolution.relative_physical_stats {
        method = method.relative_physical_stats(relative_physical_stats);
    }
    if let Some(trade_species) = evolution.trade_species {
        method = method.trade_species(trade_species.name);
    }
    if evolution.needs_overworld_rain {
        method = method.needs_overworld_rain(true);
    }
    if evolution.turn_upside_down {
        method = method.turn_upside_down(true);
    }
    if !evolution.time_of_day.is_empty() {
        method = method.time_of_day(evolution.time_of_day);
    }

    method
}

fn extract_gen_from_url(url: &str) -> Result<u8> {
    let gen_url_regex = Regex::new(r"generation/(?P<gen>\d+)/?$").unwrap();
    let caps = gen_url_regex.captures(url).unwrap();
    let generation = caps["gen"].parse::<u8>()?;
    Ok(generation)
}

fn extract_stats(stats_vec: Vec<RustemonStat>) -> Stats {
    let mut stats = Stats::default();

    for RustemonStat {
        stat, base_stat, ..
    } in stats_vec
    {
        match stat.name.as_str() {
            "hp" => stats.hp = base_stat,
            "attack" => stats.attack = base_stat,
            "defense" => stats.defense = base_stat,
            "special-attack" => stats.special_attack = base_stat,
            "special-defense" => stats.special_defense = base_stat,
            "speed" => stats.speed = base_stat,
            _ => (),
        }
    }

    stats
}

trait Past<T> {
    async fn generation(&self, api: &ApiWrapper) -> u8;
    fn value(self) -> T;
}

impl Past<Vec<RustemonTypeSlot>> for RustemonPastPokemonType {
    async fn generation(&self, _api: &ApiWrapper) -> u8 {
        extract_gen_from_url(&self.generation.url).unwrap()
    }

    fn value(self) -> Vec<RustemonTypeSlot> {
        self.types
    }
}

impl Past<RustemonTypeRelations> for RustemonPastTypeRelations {
    async fn generation(&self, _api: &ApiWrapper) -> u8 {
        extract_gen_from_url(&self.generation.url).unwrap()
    }

    fn value(self) -> RustemonTypeRelations {
        self.damage_relations
    }
}

impl Past<RustemonPastMoveStats> for RustemonPastMoveStats {
    async fn generation(&self, api: &ApiWrapper) -> u8 {
        api.get_generation(&self.version_group.name).await.unwrap() - 1
    }

    fn value(self) -> RustemonPastMoveStats {
        self
    }
}

impl Past<Vec<RustemonEffect>> for RustemonPastAbilityEffect {
    async fn generation(&self, api: &ApiWrapper) -> u8 {
        api.get_generation(&self.version_group.name).await.unwrap() - 1
    }

    fn value(self) -> Vec<RustemonEffect> {
        self.effect_entries
    }
}
