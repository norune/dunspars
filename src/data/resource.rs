use std::collections::HashMap;
use std::fs;
use std::path::PathBuf;

use anyhow::{bail, Result};

use rustemon::client::RustemonClient;

use crate::data::api::utils::{self, capture_gen_url};
use crate::data::Game;

pub enum ResourceResult {
    Valid,
    Invalid(Vec<String>),
}

pub trait Resource: Sized {
    fn get_matches(&self, value: &str) -> Vec<String> {
        self.resource()
            .iter()
            .filter_map(|r| {
                let close_enough = if !r.is_empty() && !value.is_empty() {
                    let first_r = r.chars().next().unwrap();
                    let first_value = value.chars().next().unwrap();

                    // Only perform spellcheck on first character match; potentially expensive
                    first_r == first_value && strsim::levenshtein(r, value) < 4
                } else {
                    false
                };

                if r.contains(value) || close_enough {
                    Some(r.clone())
                } else {
                    None
                }
            })
            .collect::<Vec<String>>()
    }

    fn check(&self, value: &str) -> ResourceResult {
        let matches = self.get_matches(value);
        if matches.iter().any(|m| *m == value) {
            ResourceResult::Valid
        } else {
            ResourceResult::Invalid(matches)
        }
    }

    fn validate(&self, value: &str) -> Result<String> {
        let value = value.to_lowercase();
        match self.check(&value) {
            ResourceResult::Valid => Ok(value),
            ResourceResult::Invalid(matches) => bail!(Self::invalid_message(&value, &matches)),
        }
    }

    fn invalid_message(value: &str, matches: &[String]) -> String {
        let resource_name = Self::label();
        let mut message = format!("{resource_name} '{value}' not found.");

        if matches.len() > 20 {
            message += " Potential matches found; too many to display.";
        } else if !matches.is_empty() {
            message += &format!(" Potential matches: {}.", matches.join(" "));
        }

        message
    }

    fn resource(&self) -> Vec<String>;
    fn label() -> &'static str;
}

#[derive(Debug)]
pub struct PokemonResource {
    resource: Vec<String>,
}
impl PokemonResource {
    pub async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_pokemon(client).await?;
        Ok(Self { resource })
    }
}
impl Resource for PokemonResource {
    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Pok√©mon"
    }
}

#[derive(Debug)]
pub struct TypeResource {
    resource: Vec<String>,
}
impl TypeResource {
    pub async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_types(client).await?;
        Ok(Self { resource })
    }
}
impl Resource for TypeResource {
    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Type"
    }
}

#[derive(Debug)]
pub struct MoveResource {
    resource: Vec<String>,
}
impl MoveResource {
    pub async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_moves(client).await?;
        Ok(Self { resource })
    }
}
impl Resource for MoveResource {
    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Move"
    }
}

#[derive(Debug)]
pub struct AbilityResource {
    resource: Vec<String>,
}
impl AbilityResource {
    pub async fn try_new(client: &RustemonClient) -> Result<Self> {
        let resource = utils::get_all_abilities(client).await?;
        Ok(Self { resource })
    }
}
impl Resource for AbilityResource {
    fn resource(&self) -> Vec<String> {
        self.resource.clone()
    }

    fn label() -> &'static str {
        "Ability"
    }
}

#[derive(Debug)]
pub struct GameResource {
    resource: HashMap<String, Game>,
}
impl GameResource {
    pub fn try_new() -> Result<Self> {
        let mut resource = HashMap::new();

        let data_dir = app_directory_data("resources/games.yaml");
        let game_data: Vec<Game> = serde_yaml::from_str(&fs::read_to_string(data_dir)?)?;

        for game in game_data {
            resource.insert(game.name.clone(), game);
        }

        Ok(Self { resource })
    }
}
impl Resource for GameResource {
    fn resource(&self) -> Vec<String> {
        let mut games = self.resource.iter().map(|r| r.1).collect::<Vec<&Game>>();
        games.sort_by_key(|g| g.order);

        games
            .iter()
            .map(|g| g.name.clone())
            .collect::<Vec<String>>()
    }

    fn label() -> &'static str {
        "Game"
    }
}
pub trait GetGeneration {
    fn get_gen(&self, game: &str) -> u8;
    fn get_gen_from_url(&self, url: &str) -> u8;
}

impl GetGeneration for GameResource {
    fn get_gen(&self, game: &str) -> u8 {
        self.resource.get(game).unwrap().generation
    }

    fn get_gen_from_url(&self, url: &str) -> u8 {
        capture_gen_url(url).unwrap()
    }
}

enum AppDirectories {
    Cache,
    Data,
    Config,
}

impl std::fmt::Display for AppDirectories {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            AppDirectories::Cache => write!(f, "Cache"),
            AppDirectories::Data => write!(f, "Data"),
            AppDirectories::Config => write!(f, "Config"),
        }
    }
}

pub fn app_directory_cache(target_dir: &str) -> PathBuf {
    app_directory(AppDirectories::Cache, target_dir)
}
pub fn app_directory_data(target_dir: &str) -> PathBuf {
    app_directory(AppDirectories::Data, target_dir)
}
pub fn app_directory_config(target_dir: &str) -> PathBuf {
    app_directory(AppDirectories::Config, target_dir)
}

fn app_directory(base_dir: AppDirectories, target_dir: &str) -> PathBuf {
    let base_path_buf = match base_dir {
        AppDirectories::Cache => dirs::cache_dir(),
        AppDirectories::Data => dirs::data_local_dir(),
        AppDirectories::Config => dirs::config_local_dir(),
    };
    let mut directory = base_path_buf.unwrap_or_else(|| panic!("{base_dir} directory not found"));

    directory.push(format!("dunspars/{target_dir}"));
    directory
}
