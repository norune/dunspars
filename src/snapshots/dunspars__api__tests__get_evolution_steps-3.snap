---
source: src/api.rs
expression: politoed
---
name: poliwag
methods: []
evolves_to:
  - name: poliwhirl
    methods:
      - trigger: level-up
        item: ~
        gender: ~
        held_item: ~
        known_move: ~
        known_move_type: ~
        location: ~
        min_level: 25
        min_happiness: ~
        min_beauty: ~
        min_affection: ~
        needs_overworld_rain: ~
        party_species: ~
        party_type: ~
        relative_physical_stats: ~
        time_of_day: ~
        trade_species: ~
        turn_upside_down: ~
    evolves_to:
      - name: poliwrath
        methods:
          - trigger: use-item
            item: water-stone
            gender: ~
            held_item: ~
            known_move: ~
            known_move_type: ~
            location: ~
            min_level: ~
            min_happiness: ~
            min_beauty: ~
            min_affection: ~
            needs_overworld_rain: ~
            party_species: ~
            party_type: ~
            relative_physical_stats: ~
            time_of_day: ~
            trade_species: ~
            turn_upside_down: ~
        evolves_to: []
      - name: politoed
        methods:
          - trigger: trade
            item: ~
            gender: ~
            held_item: kings-rock
            known_move: ~
            known_move_type: ~
            location: ~
            min_level: ~
            min_happiness: ~
            min_beauty: ~
            min_affection: ~
            needs_overworld_rain: ~
            party_species: ~
            party_type: ~
            relative_physical_stats: ~
            time_of_day: ~
            trade_species: ~
            turn_upside_down: ~
        evolves_to: []
